ComponentKit/Utilities/CKComponentAction.h

```diff
@@ -11,44 +11,180 @@
 #import <UIKit/UIKit.h>
 
 #import <ComponentKit/CKComponentViewAttribute.h>
+#import <ComponentKit/CKComponentActionInternal.h>
 
 @class CKComponent;
 
 /**
- A component action is simply a selector. The selector may optionally take one argument: the sending component.
+ CKTypedComponentAction is a struct that represents a method invocation that can be passed to a child component to
+ trigger a method invocation on a target.
+ 
+ We allow a typed specification of the parameters that will be provided as arguments to the component action at runtime
+ through the variadic templated arguments. You may specify an arbitrary number of arguments to your component action,
+ and you may use either object, or primitive arguments. Only trivially-constructible arguments or pointers can be used
+ as types for component actions, so primitive types like int, CGRect, and NSObject * are fine. This is enforced via a
+ compile time check.
+
+ Methods will always be provided the sender as the first argument.
+ 
+ Usage in your component header:
+ 
+ @interface MyComponent : CKComponent
+ + (instancetype)newWithAction:(CKTypedComponentAction<NSString *, int>)action;
+ @end
+ 
+ When creating the action:
+ 
+ Option 1 - Scope action. Similar to target/selector action in that it skips the responder chain from the sender, and
+            directly invokes the selector on the component or controller corresponding with the scope. Promotion is also
+            supported for scope-based actions. Scope actions weakly capture the component or controller. Does not
```

Comment:
nit: `s/Does not/They do not/` or `s/Scope actions weakly capture/A scope action weakly captures/`